src/ConfigDiscoveryTrait.php

```diff
@@ -0,0 +1,151 @@
+
+    /**
+     * @var null|string
+     */
+    private $configCacheDir;
+
+    /**
+     * @var null|string
+     */
+    private $configCacheKey;
+
+    /**
+     * Removes the application configuration cache file, if present.
+     *
+     * @param string $configCacheFile
+     */
+    private function removeConfigCacheFile($configCacheFile)
+    {
+        if (! $configCacheFile || ! file_exists($configCacheFile)) {
+            return;
+        }
+
+        unlink($configCacheFile);
+    }
+
+    /**
+     * Retrieve the config cache file, if any.
+     *
+     * @return false|string
+     */
+    private function getConfigCacheFile()
+    {
+        $configCacheDir = $this->getConfigCacheDir();
+        $configCacheKey = $this->getConfigCacheKey();
+
+        if (empty($configCacheDir)) {
+            return false;
+        }
+
+        $path = sprintf('%s/%s.', $configCacheDir, $this->configCacheBase);
+
+        if (! empty($configCacheKey)) {
+            $path .= $configCacheKey . '.';
+        }
+
+        return $path . 'php';
+    }
+
+    /**
+     * Return the configured configuration cache directory, if any.
+     *
+     * @return null|string
+     */
+    public function getConfigCacheDir()
+    {
+        if ($this->configCacheDir) {
+            return $this->configCacheDir;
+        }
+
+        $config = $this->getApplicationConfig();
+        if (isset($config['module_listener_options']['cache_dir'])) {
+            $this->configCacheDir = $config['module_listener_options']['cache_dir'];
+        }
+
+        return $this->configCacheDir;
+    }
+
+    /**
+     * Return the configured configuration cache key, if any.
+     *
+     * @return null|string
+     */
+    private function getConfigCacheKey()
+    {
+        if ($this->configCacheKey) {
+            return $this->configCacheKey;
+        }
+
+        $config = $this->getApplicationConfig();
+        if (isset($config['module_listener_options']['config_cache_key'])) {
+            $this->configCacheKey = $config['module_listener_options']['config_cache_key'];
+        }
+
+        return $this->configCacheKey;
+    }
+
+    /**
+     * Return the application configuration.
+     *
+     * Raises an exception if retrieved configuration is not an array.
+     *
+     * @return array
+     * @throws RuntimeException if config/application.config.php does not
+     *     return an array
+     */
+    function getApplicationConfig()
+    {
+        if ($this->applicationConfig) {
+            return $this->applicationConfig;
+        }
+
+        $applicationConfig = (isset($this->projectDir) && ! empty($this->projectDir))
+            ? sprintf('%s/%s', $this->projectDir, $this->applicationConfigPath)
+            : $this->applicationConfigPath;
+        if (! file_exists($applicationConfig)) {
+            $this->applicationConfig = [];
+            return $this->applicationConfig;
+        }
+
+        $config = include $applicationConfig;
```

Comment:
I think it should be:
`$this->applicationConfig = include $applicationConfig;`

variable and property name are confusing - one is file one is array...